src/vmm/src/vmm_config/drive.rs

```diff
@@ -70,109 +81,114 @@ pub struct BlockDeviceConfig {
-    }
-
-    fn get_index_of_drive_path(&self, drive_path: &PathBuf) -> Option<usize> {
-        self.config_list
+        self.list
             .iter()
-            .position(|cfg| cfg.path_on_host.eq(drive_path))
+            .position(|b| b.lock().unwrap().id().eq(drive_id))
     }
 
-    /// Inserts `block_device_config` in the block device configuration list.
-    /// If an entry with the same id already exists, it will attempt to update
-    /// the existing entry.
+    /// Inserts a `Block` in the block devices list using the specified configuration.
+    /// If a block with the same id already exists, it will overwrite it.
     /// Inserting a secondary root block device will fail.
-    pub fn insert(&mut self, block_device_config: BlockDeviceConfig) -> Result<()> {
-        // If the id of the drive already exists in the list, the operation is update.
-        match self.get_index_of_drive_id(&block_device_config.drive_id) {
-            Some(index) => self.update(index, block_device_config),
-            None => self.create(block_device_config),
-        }
-    }
-
-    fn create(&mut self, block_device_config: BlockDeviceConfig) -> Result<()> {
-        // check if the path exists
-        if !block_device_config.path_on_host.exists() {
-            return Err(DriveError::InvalidBlockDevicePath);
-        }
-
-        if self
-            .get_index_of_drive_path(&block_device_config.path_on_host)
-            .is_some()
-        {
-            return Err(DriveError::BlockDevicePathAlreadyExists);
-        }
-
-        // check whether the Device Config belongs to a root device
-        // we need to satisfy the condition by which a VMM can only have on root device
-        if block_device_config.is_root_device {
-            if self.has_root_block {
-                return Err(DriveError::RootBlockDeviceAlreadyAdded);
-            } else {
-                self.has_root_block = true;
-                // Root Device should be the first in the list whether or not PARTUUID is specified
-                // in order to avoid bugs in case of switching from partuuid boot scenarios to
-                // /dev/vda boot type.
-                self.config_list.push_front(block_device_config);
+    pub fn insert(&mut self, config: BlockDeviceConfig) -> Result<()> {
+        let is_root_device = config.is_root_device;
+        let position = self.get_index_of_drive_id(&config.drive_id);
+        let block_dev = Arc::new(Mutex::new(Self::create_block(config)?));
+        // If the id of the drive already exists in the list, the operation is update/overwrite.
+        match position {
+            // New block device.
+            None => {
+                // Check whether the Device Config belongs to a root device,
+                // we need to satisfy the condition by which a VMM can only have one root device.
+                if is_root_device {
+                    if self.has_root_block {
```

Comment:
You can rule out duplicate root device before looking at it's position.

Reply:
That's right. This is over-engineered to do everything in a single pass, but it's not really necessary. Number of block devices is never high enough to introduce any performance hit by traversing this list multiple times.

I will refactor it to try and make it simpler.